File: julia/sample3.jl
# -*- coding: utf-8 -*-

using Einsum

function benchmark(;dim1=30, dim2=20)
    A = randn(dim1, dim2)
    B = randn(dim1, dim2)
    C = randn(dim1, dim2)
    D = randn(dim1, dim2)
    E = randn(dim1, dim2)
    X = zeros(dim2, dim2, dim2, dim2, dim2)
    @einsum X[a,b,c,d,e] = A[r,a] * B[r,b] * C[r,c] * D[r,d] * E[r,d]
end


open("../time_julia.csv", "w") do fp
    for n = 5:39
        tic()
        benchmark(dim2=n)
        write(fp, "$n,$(toq())\n")
    end
end

File: python/sample3.py
# -*- coding:utf-8 -*-

import numpy as np
import functools
import timeit

def benchmark(dim1=30, dim2=20):
  A = np.random.randn(dim1, dim2)
  B = np.random.randn(dim1, dim2)
  C = np.random.randn(dim1, dim2)
  D = np.random.randn(dim1, dim2)
  E = np.random.randn(dim1, dim2)
  return np.einsum("ra,rb,rc,rd,re->abcde", A, B, C, D, E)


with open("../time_python.csv", "w") as f:
  for n in range(5, 40):
    t = timeit.Timer(functools.partial(benchmark, 30, n)).timeit(number=1)
    f.write("{},{}\n".format(n, t))
